pg_search/src/postgres/customscan/pdbscan/exec_methods/top_n.rs

```diff
@@ -140,106 +111,32 @@ impl ExecMethod for TopNScanExecState {
-                    // based on the proportion of dead tuples compared to live tuples
-                    let heaprelid = self.heaprelid;
-                    let n_dead = direct_function_call::<i64>(
-                        pg_sys::pg_stat_get_dead_tuples,
-                        &[heaprelid.into_datum()],
-                    )
-                    .unwrap();
-                    let n_live = direct_function_call::<i64>(
-                        pg_sys::pg_stat_get_live_tuples,
-                        &[heaprelid.into_datum()],
-                    )
-                    .unwrap();
-
-                    (1.0 + ((1.0 + n_dead as f64) / (1.0 + n_live as f64))).ceil() as usize
-                } else {
-                    // we've already done chunking, so just use a default scaling factor
-                    // to avoid exponentially growing the chunk size
-                    SUBSEQUENT_RETRY_SCALE_FACTOR
-                };
-
-                // set the chunk size to the scaling factor times the limit
-                self.chunk_size = (self.chunk_size * factor)
-                    .max(self.limit * factor)
-                    .min(MAX_CHUNK_SIZE);
-
-                let mut results = self.query_more_results(state, Some(self.current_segment), true);
-
-                // fast forward and stop on the ctid we last found
-                for (scored, doc_address) in &mut results {
-                    if scored.ctid == self.last_ctid {
-                        // we've now advanced to the last ctid we found
-                        break;
-                    }
+                None => {
+                    self.search_results = self.query_more_results(state);
+                    continue;
+                }
+                Some((scored, doc_address)) => {
+                    return ExecState::Ctid {
+                        ctid: scored.ctid,
+                        score: scored.bm25,
+                        doc_address,
+                    };
                 }
-
-                // this should be the next valid tuple after that, or None if `results` are now empty
-                next = results.next();
-
-                // we now have a new iterator of results to use going forward
-                self.search_results = results;
-
-                // but we'll loop back around and evaluate whatever `next` is now pointing to
-                continue;
             }
         }
     }
 
     fn reset(&mut self, _state: &mut PdbScanState) {
         // Reset tracking state but don't clear search_results
-
-        // Reset counters - excluding nresults which tracks processed results
-        self.have_less = false;
-        self.did_query = false;
-
-        // Reset the tracking state
-        self.last_ctid = 0;
-        self.found = 0;
-        self.chunk_size = 0;
-        self.retry_count = 0;
-        self.current_segment = SegmentId::default();
+        // TODO: Why?
```

Comment:
@mdashti: I don't know enough about the `reset` interface to know why we shouldn't clear the search results here.

Reply:
reset should reset the state of execution, as if you've initialized the exec method, but haven't ran the exec method. The main difference with `init` is that we want to be smart and avoid doing the query again, as the result is guaranteed to be the same. We only want to to enable iterating over the result another time.

Reply:
Of course, removing the search result will also work, but defeats the purpose of rescan, as you're doing the extra effort to search again, while you could skip it.

Reply:
Hm. I think that we must reset the `SearchResults`, from a correctness perspective. They are not exhaustive, and are created based on the contents of the `ParallelScanState`. Will add a comment. Thanks!